File: Algorithm/SegmentTree/atcoder/A58/A58.ts
import * as readline from 'readline';

class SegmentTree {
    private size: number;
    private tree: number[];

    constructor(n: number) {
        this.size = 1;
        while (this.size < n) this.size <<= 1;
        this.tree = Array(this.size * 2).fill(0);
    }

    // 更新処理: A[pos] = value
    update(pos: number, value: number): void {
        pos += this.size;
        this.tree[pos] = value;
        while (pos > 1) {
            pos >>= 1;
            this.tree[pos] = Math.max(this.tree[pos * 2], this.tree[pos * 2 + 1]);
        }
    }

    // クエリ処理: 区間 [l, r) の最大値を取得
    query(l: number, r: number): number {
        l += this.size;
        r += this.size;
        let res = 0;
        while (l < r) {
            if (l % 2 === 1) res = Math.max(res, this.tree[l++]);
            if (r % 2 === 1) res = Math.max(res, this.tree[--r]);
            l >>= 1;
            r >>= 1;
        }
        return res;
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

let n = 0,
    q = 0,
    count = 0;
const queries: string[] = [];

rl.on('line', (line) => {
    if (count === 0) {
        [n, q] = line.split(' ').map(Number);
    } else {
        queries.push(line);
        if (queries.length === q) rl.close();
    }
    count++;
});

rl.on('close', () => {
    const seg = new SegmentTree(n);
    const result: number[] = [];

    for (const query of queries) {
        const parts = query.split(' ').map(Number);
        if (parts[0] === 1) {
            const [, pos, x] = parts;
            seg.update(pos - 1, x); // 0-based
        } else {
            const [, l, r] = parts;
            result.push(seg.query(l - 1, r - 1));
        }
    }

    console.log(result.join('\n'));
});

File: Algorithm/SegmentTree/atcoder/A58/A58.py
import sys
input = sys.stdin.read

class SegmentTree:
    def __init__(self, n):
        self.size = 1
        while self.size < n:
            self.size <<= 1
        self.tree = [0] * (2 * self.size)
    
    def update(self, pos, value):
        pos += self.size
        self.tree[pos] = value
        while pos > 1:
            pos >>= 1
            self.tree[pos] = max(self.tree[2 * pos], self.tree[2 * pos + 1])
    
    def query(self, l, r):
        l += self.size
        r += self.size
        res = 0
        while l < r:
            if l % 2 == 1:
                res = max(res, self.tree[l])
                l += 1
            if r % 2 == 1:
                r -= 1
                res = max(res, self.tree[r])
            l >>= 1
            r >>= 1
        return res

def main():
    data = input().split()
    n, q = int(data[0]), int(data[1])
    seg = SegmentTree(n)
    idx = 2
    res = []
    for _ in range(q):
        c = int(data[idx])
        if c == 1:
            pos = int(data[idx + 1]) - 1  # 0-based index
            x = int(data[idx + 2])
            seg.update(pos, x)
            idx += 3
        else:
            l = int(data[idx + 1]) - 1  # 0-based
            r = int(data[idx + 2]) - 1
            res.append(str(seg.query(l, r)))
            idx += 3
    print("\n".join(res))

if __name__ == "__main__":
    main()
